rfcs/20220223-convert_the_store_from_vuex_to_pinia.md

```diff
@@ -0,0 +1,127 @@
+[@obulat's PoC PR with conversion of active-media store to a Pinia setup store]
+
+
+## Implementation plan
+
+There are two options available:
+
+1. Convert the store from Vuex to Pinia gradually, one store at a time. 
+@dhruvkb has created an excellent PoC PR for a gradual conversion to Pinia while keeping other stores in Vuex.
+
+To minimise disruption, each store can be changed using a lock-based approach where the developer takes a lock on the store using a notification to the group, makes a PR updating a single store and its usages and this lock is released after the PR is merged.
+
+This prevents PRs from diverging too much and allows a seamless transition for all involved.
+
+2. Create a feature base branch, do all the conversion in it, and only merge a fully-converted store into main. This will ensure that we are merging a fully-working converted store, and if we find any blocking concern, we can easily revert the changes by closing the feature branch.
+
+While Pinia docs suggest that it is possible to migrate large projects from Vuex to Pinia module by module, they also say that using `disableVuex: false` is not recommended. Also, I'm not sure how to use Pinia stores from Vuex modules, and we do have several interdependent modules that call one another.
+  
+## Converting a Vuex namespaced store module to a Pinia store
+There is an excellent guide on conversion in Pinia docs. This section will give steps that need to be followed for each store, and add some details specific to Openverse setup.
+
+For each store, there are several things that need to be done:
+- convert the store creation to use `defineStore`.
+- convert state to `reactive`
+- convert `getters` without params to `computed` and `getters` with parameters to functions.
+- convert `actions` and `mutations` to functions. They will no longer use the `context` or `state` parameter, and will, instead of using an all-capital constant for a name, be renamed to use camelCase (e.g. `[FETCH_MEDIA]` will become `fetchMedia`).
+- return the necessary state properties, getters and actions.
+- adjust the types where necessary.
+- remove the unused constants from - `~/constants/mutation-types.js`, `~/constants/action-types.js`, `~/constants/store-modules.js`.
+- refactor the use of the store in components.
+- add or refactor the store unit test.
+- refactor the unit tests for components that use this store.
+
+### State
+When using the setup store, we should try to use a `reactive` to create the initial state object, and not a set of `ref`s. Otherwise, we can create errors by accidentally having a state property name and an action parameter that is updating it with the same name. This doesn't work:
+```js
+const message = ref('string message')
+function setMessage({message}){
+    message.value = message
+}
+```
+Here, the destructured parameter name `message` shadows the state property name `message`, and the `setMessage` doesn't update the state as could be expected.
```

Comment:
:scream: This makes me really want to enable that `no-shadow-vars` rule in ESLint